afkak/test/test_consumer.py

```diff
@@ -1756,8 +1757,59 @@ def test_consumer_process_messages_should_exit_when_no_messages_left(self):
         consumer = Consumer(
             client, 'a_topic', a_partition, a_processor, a_consumer_group,
         )
+        self.assertIsNone(self.successResultOf(consumer._process_messages([])))
 
-        self.assertIsNone(consumer._process_messages([]))
+    def test_consumer_process_messages_stack_safe(self):
+        """
+        The Consumer process a list of messages without causing a RecursionError
+        """
+
+        count = sys.getrecursionlimit() + 1
+
+        class MessageProcessor():
+            calls = 0
+
+            def __call__(self, consumer, messages):
+                self.calls += 1
+                return succeed(None)
+
+        # set up a Consumer and start it so it can consume messages
+        fetch_ds = Deferred()
+
+        offset = 0
+        topic = 'some_topic'
+        partition = 13
+        consumer_group = "A nice day"
+
+        clock = MemoryReactorClock()
+        mockclient = Mock(reactor=clock)
+        mockclient.send_fetch_request.side_effect = [fetch_ds]
+
+        responses = [FetchResponse(topic, partition, KAFKA_SUCCESS, 5, [])]
+        processor = MessageProcessor()
+        consumer = Consumer(
+            mockclient, topic, partition, processor, consumer_group, auto_commit_every_n=1)
+        consumer.start(offset)
+        fetch_ds.callback(responses)
+        # At this point consumer should be started and ready to process messages
+
+        message_list = [create_message(b'paylod', b'key') for _ in range(count)]
+        message_set = KafkaCodec._encode_message_set(message_list, offset)
+        message_iter = KafkaCodec._decode_message_set_iter(message_set)
+
+        messages = []
+        for msg in message_iter:
+            messages.append(
+                SourcedMessage(
+                    message=msg.message,
+                    offset=msg.offset,
+                    topic=topic,
+                    partition=partition))
+
+        consumer._process_messages(messages)
```

Comment:
Because of how other tests are constructed I could not figure out what was the acceptable pattern of how to test functionality. In this case, I decided to go the smallest unit of functionality to test, which is whatever is contained in `_process_messages`.  I agree on is that this function is private and that having tests that target these specifically make it hard to refactor code later. On the other hand, testing the specific function, makes it less susceptible to changes or bugs introduced by changes in other areas (which integration tests are meant to catch). In reality most of the unit tests that we have here are actually integration tests (and our integration tests are really system test), and that probably makes it confusing.  I don't disagree with your suggestion, I am just making a statement trying to put myself in the shoes of other contributors.

Reply:
Yeah, I'm not faulting you for confusion, but in this case I think that it is easy to avoid poking the private state.

While I don't want to get into a discussion of what is a "unit test" vs. an "integration test", I would actually like to rework these tests to be more integration-y in the sense that I would rather they use a real `KafkaClient` that wraps a fake-I/O [`Connections`](https://github.com/ciena/afkak/blob/e46e75bd998b6d2d87512368a58145265532de0a/afkak/test/endpoints.py#L75) instance. I've done this for many tests in the `afkak.test.test_client` module, but I was trying to minimize the number of lines I touched so the result is still quite messy:

https://github.com/ciena/afkak/blob/e46e75bd998b6d2d87512368a58145265532de0a/afkak/test/test_client.py#L1147-L1234

I didn't want to ask you to make such radical changes here, but sticking to public interfaces will make the transition easier.